oemof/solph/custom.py

```diff
@@ -839,3 +844,491 @@ def tes_ub_constr_rule(block, n, t):
+
+        interval in between which total DSM  must be fully compensated (default = 24h )
+    **delay_time: int
+        only used in "delay":
+
+        the time after which the load shift must be fully compensated
+
+    Notes
+    -----
+
+    The following sets, variables, constraints and objective parts are created
+    :py:class:`~oemof.solph.custom.SinkDSMIntervalBlock` or :py:class:`~oemof.solph.custom.SinkDSMDelayBlock`
+
+    * This component is still under development.
+
+    * The interval always starts at the first timestep = 0.
+
+    * delay method based on the formulation given in a paper by Zerrahn, Alexander and Schill, Wolf-Peter (2015):
+      On the representation of demand-side management in power system models,
+      in: Energy (84), pp. 840-845, 10.1016/j.energy.2015.03.037, accessed 17.09.2019, pp. 842-843.
+
+    * in 'delay' method: delay time might be extended artificially if DSM-capacity isnt fully used.
+      see oemof/Issue: #622 for more information
+
+    """
+
+    def __init__(self, demand, c_up, c_do, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.c_up = sequence(c_up)
+        self.c_do = sequence(c_do)
+        self.demand = sequence(demand)
+        self.method = kwargs.get('method', 'delay')
+        self.shift_interval = kwargs.get('shift_interval', 24)
+        self.delay_time = kwargs.get('delay_time', 3)
+
+    def constraint_group(self):
+        possible_methods = ['delay', 'interval']
+
+        if self.method == possible_methods[0]:
+            return SinkDSMDelayBlock
+        elif self.method == possible_methods[1]:
+            return SinkDSMIntervalBlock
+        else:
+            raise ValueError('The method selection must be one of the following set: '
+                             '"{}"'.format('","'.join(possible_methods)))
+
+#######################################################################################
+#                      Interval Method
+
+
+class SinkDSMIntervalBlock(SimpleBlock):
+    r"""Block for the linear relation of a DSM component and an electrical bus using 'interval' method.
+    Load shift must be compensated during a specific interval, starting at timestep 0 (default=24h).
+
+
+    **The following constraints are created for method = 'potential':**
+
+    .. _SinkDSMInterval-equations:
+
+    .. math::
+        &
+        (1) \quad  C_{t}^{do} < DSM_{t}^{updown} < C_{t}^{up} \quad \forall t \\
+        &
+        (2) \quad flow = demand_{t} + DSM_{t}^{updown} \quad \forall t \\
+        &
+        (3) \quad  \sum_{t=0}^{24} DSM_{t}^{updown} = 0 \quad \forall t \\
+        &
+
+
+
+    **Table: Symbols and attribute names of variables and parameters**
+
+        .. csv-table:: Variables (V) and Parameters (P)
+            :header: "symbol", "attribute", "type", "explanation"
+            :widths: 1, 1, 1, 1
+
+            ":math:`DSM_{t}^{updown}` ",":py:obj:`DSMupdown[g, tt]` ", "V", "DSM up/down shift"
```

Comment:
It is not immediately clear, what the variables/parameters mean. https://oemof.readthedocs.io/en/latest/api/oemof.solph.html#oemof.solph.components.GenericStorageBlock can be used as a best-practice example:

* For the storage, we use E for energies are call them capacity.
* Normalized units are "levels" and have the symbol "c".
* Flows are \dot{E}.